geoopt/optim/rwolfe.py

```diff
@@ -0,0 +1,249 @@
+"""Riemannian Line Search.
+
+This module implements line search on Riemannian manifolds using geoopt.
+This module uses the same syntax as a Torch optimizer
+"""
+
+from .mixin import OptimMixin
+from ..tensor import ManifoldParameter, ManifoldTensor
+from scipy.optimize.linesearch import scalar_search_wolfe1
+import torch
+
+__all__ = ["RiemannianLineSearch"]
+
+
+class RiemannianLineSearch(OptimMixin, torch.optim.Optimizer):
+    r"""Riemannian line search optimizer using strong Wolfe conditions.
+
+    If we try to minimize objective $f:M\to \mathbb{R}$, then we take a step in
+    the direction $\eta=\mathrm{grad} f(x)$. We define objective function
+    $$\phi(\alpha) = f(R_x(\alpha\eta))$$, where $R_x$ is the retraction at $x$.
+    The Wolfe conditions for the step size $\alpha$ are then given by
+    $$f(R_x(\alpha\eta))\leq f(x)+c_1 \alpha \langle\mathrm{grad} f(x),\eta\rangle$$
+    and
+    $$\langle\mathrm{grad}f(R_x(\alpha\eta)),\mathcal T(\alpha \eta,\grad f(x))\rangle
+    \geq c_2\langle\mathrm{grad}f(x),\eta\rangle,$$
+    where $\mathcal T$ is the vector transport. In terms of the line search objective
+    these are simply given by
+    $$\phi(\alpha)\leq \phi(0)+c_1\alpha\phi'(0)$$
+    and
+    $$\phi'(\alpha)\geq c_2\phi'(0)$$
+
+    The constants $c_1$ and $c_2$ satisfy $c_1\in (0,1)$ and $c_2\in (c_1,1)$.
+
+    Parameters
+    ----------
+    params : iterable
+        iterable of parameters to optimize or dicts defining
+        parameter groups
+    c1 : float
+        Parameter controlling Armijo rule (default: 1e-4)
+    c2 : float
+        Parameter controlling curvature rule (default: 0.9)
+    fallback_stepsize : float
+        fallback_stepsize to take if no point can be found satisfying the
+        Wolfe conditions (default: 1)
+    amax : float
+        maximum step size (default: 50)
+    amin : float
+        minimum step size (default: 1e-8)
+    stabilize : int
+        Stabilize parameters if they are off-manifold due to numerical
+        reasons every ``stabilize`` steps (default: ``None`` -- no stabilize)
+    """
+
+    def __init__(
+        self,
+        params,
+        stabilize=None,
+        c1=1e-4,
+        c2=0.9,
+        fallback_stepsize=1,
+        amax=50,
+        amin=1e-8,
+    ):
+        defaults = dict(
+            c1=c1,
+            c2=c2,
+            fallback_stepsize=fallback_stepsize,
+            amax=amax,
+            amin=amin,
+            stabilize=stabilize,
+        )
+        super(RiemannianLineSearch, self).__init__(
+            params, defaults, stabilize=stabilize
+        )
+        self._params = self.param_groups[0]["params"]
+        self.c1 = self.param_groups[0]["c1"]
```

Comment:
why is it global? Is it possible to have different constants between param groups?

Reply:
`stabilize` should probably be per parameter group, but I don't see a way to implement different line search constants for different parameter groups.

Reply:
> `stabilize` should probably be per parameter group

Fair point, I missed that

Reply:
> I don't see a way to implement different line search

That means to avoid misunderstanding between user and developer we need some checks for that or do not set is per parameter group in the first place